File: sword_for_offer/codes/cpp/sfo_52_the_first_common_node_in_two_linked_lists_s1/sfo_52_the_first_common_node_in_two_linked_lists_s1.cpp
/*
* File: sfo_52_the_first_common_node_in_two_linked_lists_s1.cpp
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

#include "../include/include.hpp"

// ===== Solution Code =====
class Solution {
public:
    ListNode* getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode *A = headA, *B = headB;
        while (A != B) {
            A = A != nullptr ? A->next : headB;
            B = B != nullptr ? B->next : headA;
        }
        return A;
    }
};

int main() {
    // ======= Test Case =======
    // Build two linked lists with intersection of 8
    // headA = 4 -> 1 -> 8 -> 4 -> 5
    //                   ↑
    // headB = 5 -> 0 -> 1
    ListNode* headA = vectorToLinkedList(vector<int> { 4, 1, 8, 4, 5 });
    ListNode* headB = vectorToLinkedList(vector<int> { 5, 0, 1, 8, 4, 5 });
    ListNode* intersectA = getListNode(headA, 8);
    ListNode* intersectB = getListNode(headB, 1);
    intersectB->next = intersectA; // Concatenate the two lists
    // ====== Driver Code ======
    Solution* slt = new Solution();
    ListNode* res = slt->getIntersectionNode(headA, headB);
    cout << res->val << endl;

    return 0;
}

File: sword_for_offer/codes/python/sfo_52_the_first_common_node_in_two_linked_lists_s1.py
'''
File: sfo_52_the_first_common_node_in_two_linked_lists_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        A, B = headA, headB
        while A != B:
            A = A.next if A else headB
            B = B.next if B else headA
        return A

# ======= Test Case =======
# Build two linked lists with intersection of 8
# headA = 4 -> 1 -> 8 -> 4 -> 5
#                   ↑
# headB = 5 -> 0 -> 1
headA = list_to_linked_list([4, 1, 8, 4, 5])
headB = list_to_linked_list([5, 0, 1, 8, 4, 5])
intersectA = get_list_node(head=headA, val=8)
intersectB = get_list_node(head=headB, val=1)
intersectB.next = intersectA # Concatenate the two lists
# ====== Driver Code ======
slt = Solution()
res = slt.getIntersectionNode(headA, headB)
print(res.val)
